st3/lsp_utils/v1.py

```diff
@@ -0,0 +1,143 @@
+import shutil
+import sublime
+import weakref
+import os
+import subprocess
+from sublime_lib import ResourcePath
+from LSP.plugin import AbstractPlugin
+from LSP.plugin import ClientConfig
+from LSP.plugin import Response
+from LSP.plugin import Session
+from LSP.plugin import WorkspaceFolder
+from LSP.plugin.core.rpc import method2attr  # TODO: Remove this import
+from LSP.plugin.core.typing import List, Optional, Dict, Callable, Any
+
+
+def is_node_installed():
+    return shutil.which('node') is not None
+
+
+class ApiWrapper:
+    def __init__(self, plugin: AbstractPlugin):
+        self.__plugin = plugin
+
+    def on_notification(self, method: str, handler: Callable) -> None:
+        setattr(self.__plugin, method2attr(method), handler)
+
+    def on_request(self, method: str, handler: Callable) -> None:
+
+        def send_response(request_id, result):
+            session = self.__plugin.weaksession()
+            if session:
+                session.send_response(Response(request_id, result))
+
+        def on_response(params, request_id):
+            handler(params, lambda result: send_response(request_id, result))
+
+        setattr(self.__plugin, method2attr(method), on_response)
+
+
+class NpmClientHandler(AbstractPlugin):
+    # To be overridden by subclass.
+    package_name = None  # type: Optional[str]
+    server_directory = None  # type: Optional[str]
+    server_binary_path = None  # type: Optional[str]
+    # Internal
+    _package_cache_path = None  # type: Optional[str]
+    _cache_server_path = None  # type: Optional[str]
+
+    @classmethod
+    def setup(cls) -> None:
+        assert cls.package_name
+        assert cls.server_directory
+        assert cls.server_binary_path
+        if cls._package_cache_path:
+            return
+        cls._package_cache_path = os.path.join(sublime.cache_path(), cls.package_name)
+        cls._cache_server_path = os.path.join(cls._package_cache_path, cls.server_directory)
+
+    @classmethod
+    def cleanup(cls) -> None:
+        cls._package_cache_path = None
+        cls._cache_server_path = None
+
+    @classmethod
+    def get_binary_arguments(cls):
+        """
+        Returns a list of extra arguments to append when starting server.
+        """
+        return ['--stdio']
+
+    @classmethod
+    def name(cls) -> str:
+        assert cls.package_name
+        return cls.package_name[len('LSP-'):]
+
+    @classmethod
+    def configuration(cls) -> sublime.Settings:
+        if not cls._package_cache_path:
+            # https://github.com/sublimelsp/LSP/issues/899
+            cls.setup()
+        settings = super().configuration()
```

Comment:
It might be a bit annoying to have a `sublime.Settings` object and not a Dict here.

I suppose this is to allow mutating settings but in that case, how would the changes be saved? Would this package have to "guess" settings file name? I mean, it's not hard to guess but it's not explicitly defined here either.

Maybe there should be a separate method (like before) that would get the initial `sublime.Settings` and could mutate it and then the final configuration would be received here as a Dict?

The way it is now also forced you to do the ugly hack with calling `on_client_configuration_ready` twice.

Reply:
I'm convinced that getting a `sublime.Settings` object from a plugin is good, because the base LSP package can then listen for changes (add_on_change, clear_on_change). When a change occurs, we can hot-reload in LSP. At least, that's the goal.

> The way it is now also forced you to do the ugly hack with calling on_client_configuration_ready twice.

Well, eventually `on_client_configuration_ready` will disappear :)

> how would the changes be saved? 

Hmm, I don't understand why that would be needed? Why do we need to save changes to the sublime.Settings object to disk?

It's only for listening for changes from the user. If the user saves their overrides in Packages/User/LSP-foo.sublime-settings, the on_change callback is called.

Reply:
> Hmm, I don't understand why that would be needed? Why do we need to save changes to the sublime.Settings object to disk?

We've done that before to migrate to new settings structure (the code is still there in the old handler).

And we might want to migrate user settings when some server does a breaking change that requires us to update user's settings.

I think my suggestion for having two separate endpoint is still worth it.

Reply:
Well, saving is quite simple with `sublime.save_settings('LSP-{}.sublime-settings'.format(cls.name()))`. And server settings can be obtained with `settings.get("settings")`. One can do the migration in the `def configuration` function. I don't see the problem?

Reply:
- The name of the settings file is not explicit so have to be "guessed" now. Granted, it's not hard to guess but still. Maybe it should be explicitly stated by the plugin like before?
- Otherwise I don't have strong arguments against it.

Reply:
> The name of the settings file is not explicit so have to be "guessed" now.

You're right, it makes sense to do this a bit more explicit.